File: python3/can_place_flowers.py
"""
Space   : O(1)
Time    : O(n)
"""

class Solution:
    def canPlaceFlowers(self, flowerbed: List[int], n: int) -> bool:
        count, place = 0, 0
        all_zero = True
        beginning = flowerbed[0] == 0
        leng = len(flowerbed)

        if n == 0:
            return True

        for i in range(leng):
            if flowerbed[i] == 0:
                count += 1
            else:
                all_zero = False
                if beginning:
                    place += count // 2
                    beginning = False
                elif count > 0:
                    place += (count - 1) // 2
                count = 0

        if all_zero:
            return (leng+1) // 2 >= n
                
        if count > 0:
            place += count // 2

        return place >= n

File: golang/can_place_flowers.go
package main

func canPlaceFlowers(flowerbed []int, n int) bool {
	if n == 0 {
		return true
	}

	f := len(flowerbed)

	begin, allZero := flowerbed[0] == 0, true
	count, place := 0, 0

	for i := 0; i < f; i++ {
		if flowerbed[i] == 0 {
			count++
		} else {
			allZero = false
			if begin {
				place += count / 2
				begin = false
			} else if count > 0 {
				place += (count - 1) / 2
			}
			count = 0
		}

	}

	// if all items 0
	if allZero {
		return (f+1)/2 >= n
	}

	// if end exists
	if count > 0 {
		place += count / 2
	}

	if n <= place {
		return true
	}

	return false
}
